File: 链表/234.回文链表.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        stack = []
        i, j = head, head
        while j and j.next:
            stack.append(i)
            i = i.next
            j = j.next.next
        if j:
            i = i.next
        while stack:
            cur = stack.pop()
            if cur.val != i.val:
                return False
            i = i.next
        return True

File: 链表/234.回文链表.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function (head) {
  const mid = middleNode(head);
  let head2 = reverseList(mid);
  while (head2 !== null) {
    if (head.val !== head2.val) {
      // 不是回文链表
      return false;
    }
    head = head.next;
    head2 = head2.next;
  }
  return true;
};

// 876. 链表的中间结点
function middleNode(head) {
  let slow = head,
    fast = head;
  while (fast !== null && fast.next !== null) {
    slow = slow.next;
    fast = fast.next.next;
  }
  return slow;
}

// 206. 反转链表
function reverseList(head) {
  let pre = null,
    cur = head;
  while (cur !== null) {
    const nxt = cur.next;
    cur.next = pre;
    pre = cur;
    cur = nxt;
  }
  return pre;
}
